File: C++/迭代与递归/linked.cpp
/*初始化链表 1 -> 3 -> 2 -> 5 -> 4*/
// 初始化各个节点
ListNode* n0 = new ListNode(1);
ListNode* n1 = new ListNode(3);
ListNode* n2 = new ListNode(2);
ListNode* n3 = new ListNode(5);
ListNode* n4 = new ListNode(4);
// 构建节点间引用
n0->next = n1;
n1->next = n2;
n2->next = n3;
n3->next = n4;

/*插入节点*/
void insert(ListNode *n0, ListNode *P) {
    ListNode *n1 = n0->next;
    P->next = n1;
    n0->next = P;
}

/*删除节点*/
void remove(ListNode *n0) {
    if (n0->next == nullptr) {
        return;
    }
    ListNode *P = n0->next;
    ListNode *n1 = P->next;
    n0->next = n1;
    delete P;
}

/*访问节点*/
ListNode *access(ListNode *head, int index) {
    for (int i = 0; i < index; i++) {
        if (head == nullptr)
            return nullptr;
        head = head->next;
    }
    return head;
}

/*查找结点*/
int find(ListNode *head, int target) {
    int index = 0;
    while (head != nullptr) {
        if (head->val == target) 
            return index;
        head = head->next;
        index++;
    }
    return -1;
}

File: python/迭代与递归/linked.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# 初始化链表 1 -> 3 -> 2 -> 5 -> 4
## 初始化各个节点
n0 = ListNode(1) 
n1 = ListNode(3)
n2 = ListNode(2)
n3 = ListNode(5)
n4 = ListNode(4)
## 构建节点之间的引用
n0.next = n1
n1.next = n2
n2.next = n3
n3.next = n4

# 插入节点
def insert(n0, P):
    n1 = n0.next
    p.next = n1
    n0.next = P

# 删除节点
def remove(n0):
    if not n0.next:
        return 
    P = n0.next
    n1 = P.next
    n0.next = n1

# 访问节点
def access(head,index):
    for _ in range(index):
        if not head:
            return None
        head = head.next
    return head

# 查找节点
def find(head, target):
    index = 0
    while head:
        if head.val == target:
            return index
        head = head.next
        index += 1
    return -1

